impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java

```diff
@@ -458,17 +449,32 @@ public float getHitRate(TimeUnit unit) {
         enforceCapacity(1);
         try {
           ValueHolder<V> value = fault.get();
-          OnHeapValueHolder<V> newValue = value == null ? null : newCreateValueHolder(key, value.value(), now);
-
-          if (value == null) {
+          final OnHeapValueHolder<V> newValue;
+          if(value != null) {
+            if (valueSerializer != null) {
+              newValue = new ByValueOnHeapValueHolder<V>(value, valueSerializer);
+            } else {
+              newValue = new ByRefOnHeapValueHolder<V>(value);
+            }
+            final Duration expiryForAccess = expiry.getExpiryForAccess(key, value.value());
+            if(expiryForAccess != null) {
+              newValue.setExpirationTime(expiryForAccess.getAmount(), expiryForAccess.getTimeUnit());
+            }
+          } else {
             backEnd.remove(key, fault);
             return null;
-          } else if (backEnd.replace(key, fault, newValue)) {
+          }
+
+          if (backEnd.replace(key, fault, newValue)) {
             return getValue(newValue);
           } else {
             ValueHolder<V> p = getValue(backEnd.remove(key));
             if (p != null) {
-              invalidationListener.onInvalidation(key, p);
+              if (p.isExpired(timeSource.getTimeMillis(), TimeUnit.MILLISECONDS)) {
+                eventListener.onExpiration(key, p);
+              } else {
+                eventListener.onEviction(key, p);
+              }
```

Comment:
I am not sure what this invalidation is here ... 
 . If we couldn't replace the fault with the faulted value
 . Remove what was in the `CachingTier`
 . If something was actually removed, flush it back... 

Now, how would that happen? The fault gets evicted, then a new value is installed, is that it?